src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,307 @@
+        throw new Error(
+          "VideoThumbnailLoaderError: Couldn't find thumbnail.");
+      }
+
+      log.debug("VTL: Found thumbnails for time", payload.time, thumbnails);
+
+      return { thumbnails, payload };
+    }),
+    distinctUntilChanged((a, b) => {
+      if (a.thumbnails.length !== b.thumbnails.length) {
+        return false;
+      }
+      for (let i = 0; i < a.thumbnails.length; i++) {
+        if (a.thumbnails[i].start !== b.thumbnails[i].start ||
+            a.thumbnails[i].duration !== b.thumbnails[i].duration ||
+            a.thumbnails[i].mediaURL !== b.thumbnails[i].mediaURL) {
+          return false;
+        }
+      }
+      return true;
+    }),
+    switchMap(({ thumbnails, payload: { time, resolve, reject } }) => {
+      return videoSourceInfos$.pipe(
+        mergeMap((videoSourceBuffer) => {
+          return videoSourceBuffer.removeBuffer(time - 5, time + 5).pipe(
+            mergeMap(() => {
+              if (!this._thumbnailTrack) {
+                throw new Error(
+                  "VideoThumbnailLoaderError: No thumbnail track given.");
+              }
+
+              log.debug("VTL: Removed buffer before appending segments.", time);
+
+              return getSegmentsData(thumbnails, videoElement).pipe(
+                mergeMap((data) => {
+                  if (data) {
+                    const appendBuffer$ = videoSourceBuffer
+                      .appendSegment({
+                        chunk: data,
+                        initSegment: null,
+                        codec: this._thumbnailTrack.codec });
+                    return appendBuffer$.pipe(
+                      tap(() => {
+                        log.debug("VTL: Appended segment.", data, time);
+                        this._thumbnailVideoElement.currentTime = time;
+                      })
+                    );
+                  }
+                  return observableOf(null);
+                }),
+                finalize(() => {
+                  resolve();
+                })
+              );
+            }));
+          }),
+          catchError((err) => {
+            reject(err);
+            return EMPTY;
+          })
+        );
+      }),
+      catchError((err) => {
+        this.dispose();
+        throw err;
+      })
+    ).subscribe();
+  }
+
+  /**
+   * Set time of thumbnail video media element :
+   * - Remove buffer when too much buffered data
+   * - Search for thumbnail track element to display
+   * - Load data
+   * - Append data
+   * Resolves when time is set.
+   * @param {number} time
+   * @returns {Promise}
+   */
+  setTime(time: number) {
+    return new Promise((resolve, reject) => {
+      this._setTime$.next({ time, resolve, reject });
+    });
+  }
+
+  /**
+   * Update thumbnail track from adaptation
+   * @param {Object} adaptation
+   * @returns {Object}
+   */
+  public updateThumbnailTrack(trickModeTrack: Representation): IThumbnailTrack {
```

Comment:
What is the point to have both the constructor and that function to set the trickModeTrack.
The API may be easier to understand if we had only one or the other